test/fixtures/gcs-fixture/src/main/java/fixture/gcs/GoogleCloudStorageHttpHandler.java

```diff
@@ -129,35 +129,25 @@ public void handle(final HttpExchange exchange) throws IOException {
                 final MockGcsBlobStore.BlobVersion blob = mockGcsBlobStore.getBlob(path, ifGenerationMatch);
                 if (blob != null) {
                     final String rangeHeader = exchange.getRequestHeaders().getFirst("Range");
-                    final long offset;
-                    final long end;
+                    final BytesReference response;
                     if (rangeHeader == null) {
-                        offset = 0L;
-                        end = blob.contents().length() - 1;
+                        response = blob.contents();
                     } else {
                         final HttpHeaderParser.Range range = HttpHeaderParser.parseRangeHeader(rangeHeader);
                         if (range == null) {
                             throw new AssertionError("Range bytes header does not match expected format: " + rangeHeader);
                         }
-                        offset = range.start();
-                        end = range.end();
-                    }
 
-                    if (offset >= blob.contents().length()) {
-                        exchange.getResponseHeaders().add("Content-Type", "application/octet-stream");
-                        exchange.sendResponseHeaders(RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus(), -1);
-                        return;
-                    }
+                        if (range.start() >= blob.contents().length()) {
```

Comment:
I wonder, should we be checking the range end here? If you ask for bytes 0-10 of an 8-byte blob, we cannot satisfy that range.

Reply:
Yes, probably. I'll have a look at what the spec says, currently we just send as much of it as we have.

Reply:
I'm also adding ETags, because it looks like that's the mechanism that the `BlobReadChannel` uses to detect that the file it's streaming has changed under it. We can implement them trivially using generation I think.

Reply:
So it looks like
- We don't yet check that the file hasn't changed when resuming a download (I raised [ES-11432](https://elasticco.atlassian.net/browse/ES-11432) to address at some point)
- We do need to allow ranges beyond the end of actual content (streaming uses this to request a chunk). There may be more to the story, but if we start enforcing the ranges all that breaks.

Reply:
I confirmed this testing against the actual API
- If the start of the range is past the end of the file, you get a 416
- If the end of the range is past the end of the file, but the start is in it, you get a 206 and as much of the file as can be delivered